### ISSUE TITLE
Feature: Reorder Workspaces

### ISSUE THREAD

**Issue by CPickens42** [ferdium/ferdium-app#146]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/ferdium/ferdium-app/blob/develop/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/ferdium/ferdium-app/blob/develop/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/ferdium/ferdium-app/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

Hello, I'm always a bit frustrated when I setup a new workspace. For instance, I have a workspace with slack, discord, etc. I just setup a new workspace with my emails account. I would like it to be the 1st workspace of ferdium, but I can't reorder that at all.

### Proposed Solution

A drag and drop feature would be great.

### Alternatives Considered

Another solution would be to use Up and Down arrows icon to reorder the workspaces. Maybe it's easier to implement.

### Current workaround
So far, to reorder workspaces, I need to edit manually the sqlite database with sqlitebrowser and change manually the ids of everything ^^.

**Comment by CPickens42** [comment]
So far, to reorder workspaces and services, I need to edit manually the sqlite database with sqlitebrowser and change manually the ids of everything ^^.

**Comment by vraravam** [comment]
Reordering of services is possible by dragging and dropping within the sidebar - both when the "All workspaces" is active, or any individual workspace is active. Reordering of workspaces themselves is not yet implemented in Ferdium nor Franz afaik.

**Comment by vraravam** [comment]
Also, this fix was merged into Ferdi (from which Ferdium was born) in this commit: https://github.com/ferdium/ferdium-app/commit/17f3a22e4193dfcf8dfd5cbd5aa2e812f2cbd6ae

**Comment by vraravam** [comment]
@CPickens42 - if you are able to verify the reordering by drag-and-drop of the services, please update here and change the title of the feature enhancement - so that its less confusing for folks who come to this ticket. It will also help to have a single/small change that can be implemented, as opposed to the noise about service-reordering

**Comment by CPickens42** [comment]
@vraravam, okay sorry, you are right! I can reorder services (within a workspace)!

So, in that case, I just want to be able to reorder workspaces ! :wink:

**Comment by jeroni** [comment]
Reordering is buggy, it seems an index bug.

Version: 6.0.1-nightly
OS: Arch linux
DE: Gnome 42.3

https://user-images.githubusercontent.com/9005992/180392510-53d00e4b-46f2-4ca4-9d80-cee45bc64dac.mp4

**Comment by vraravam** [comment]
@jeroni - could you please search in the closed tickets? there's a more specific one for reordering of services. This one is for reordering of workspaces.

**Comment by 3gmgithub** [comment]
I am using Ferdium on Ubuntu 22.04 with v6.7.0 'No Server'
The reorder seems to work on the 'All Workspaces' but not on the individual workspaces....
I can drag the icons and drop them where I want them in 'All workspaces' but in say the 'Work' workspace the drag and drop does drag... But does not stay when dropped.

Am I doing this wrong?

**Comment by KelSolaar** [comment]
This is issue is really confusing and should be renamed and clarified. #1611 was close even though the issue described was really clear: Services cannot be reordered when an individual workspace is selected. I'm having the exact same problem with

```
Version: 6.7.1
Server: No Server
Electron: 29.1.1
Chrome: 122.0.6261.111
Node.js: 20.9.0
Platform: macOS Unknown
Arch: arm64
Build date: Sun Mar 10 2024 15:30:21 
```

but I'm not sure if this is the correct issue to report that as this one seems to be about reordering workspaces themselves and not the services within the workspace.

**Comment by SpecialAro** [comment]
@KelSolaar you are absolutely right. I misinterpreted issue #1611, but it is indeed a different issue. I'm reopening it back again even though I'm not able to reproduce the issue (it works for me without any issue).

**Comment by theAkito** [comment]
From how I personally use stuff like this, this is an absolutely essential feature and am wondering, how difficult this would be to implement and how long it would take. Otherwise, I need to write some Nushell script, which automatically edits the SQLite database or something.

How does anyone use workspaces without being able to re-order them? 😆

**Comment by gabrielevinci** [comment]
Hi everyone, this is literally my first post on Github, so I apologize if I will make some mistakes. Looking through perplexity I sensed that to change the workpace's order there would be a method, that is to modify a certain "Ferdium-Strage.db" which is found in "C: \ Users \ User \ AppataData \ Roaming \ Ferdium", is it correct? Has anyone written a detailed guide of how to do Step-by-step? I was unable to find it.
Thanks to all those who will help me

**Comment by theAkito** [comment]
> Hi everyone, this is literally my first post on Github, so I apologize if I will make some mistakes. Looking through perplexity I sensed that to change the workpace's order there would be a method, that is to modify a certain "Ferdium-Strage.db" which is found in "C: \ Users \ User \ AppataData \ Roaming \ Ferdium", is it correct? Has anyone written a detailed guide of how to do Step-by-step? I was unable to find it. Thanks to all those who will help me

I might have a look at it and contribute my findings to this issue, in order to get this to progress. 🙂

**Comment by gabrielevinci** [comment]
> > Hi everyone, this is literally my first post on Github, so I apologize if I will make some mistakes. Looking through perplexity I sensed that to change the workpace's order there would be a method, that is to modify a certain "Ferdium-Strage.db" which is found in "C: \ Users \ User \ AppataData \ Roaming \ Ferdium", is it correct? Has anyone written a detailed guide of how to do Step-by-step? I was unable to find it. Thanks to all those who will help me
> 
> I might have a look at it and contribute my findings to this issue, in order to get this to progress. 🙂

That would be great. I will follow this threads with anxiety

**Comment by theAkito** [comment]
I've created a proof of concept for re-ordering services within workspaces, by reverse-engineering the database structure of Ferdium.

https://gist.github.com/theAkito/daf71b45e02ecde502d55af4df0a6721

This is a [Nu](https://www.nushell.sh/) script, which should work on any client machine running Ferdium on Windows or Linux.

First, close all Ferdium processes.

Second, all you have to provide as an argument to the script is the correct display name of your workspace, including correct case, etc.

##### Example
```bash
nu ferdium-service-order-sqlite.nu Personal # Where "Personal" is the name of your workspace.
```

After running this script, the first service in your workspace will switch with the second. If you run the script one more time, they will switch back, which represents the original state of your service order.

The script is safe to use, as it always backs up the current SQLite database in use by Ferdium, before modifying anything.

@gabrielevinci

If you give me a list of your services within a workspace, in the correct order & the workspace's display name, I can give you a brief guide on how to run the script, which automatically orders the services as you need it.

**Comment by gabrielevinci** [comment]
Hello tahnks so much for the answer and for your work.
Since, unfortunately, I am still learning in the world of programming I write to you step by step what I have done for now:
- I installed Nushell with this command (Winget Install Nushell) in my cmd
- Then I installed the "vscode-nushell-lang" extension on Visual Studio Code
For now is this correct?

Also send you a screen of the names of my workspace and strengthened according to the order I would like

![Image](https://github.com/user-attachments/assets/4d8ed16d-1a37-4597-8cd6-13ea3a9c091d)

Thanks again for your work

**Comment by theAkito** [comment]
> Hello tahnks so much for the answer and for your work. Since, unfortunately, I am still learning in the world of programming I write to you step by step what I have done for now:
> 
>     * I installed Nushell with this command (Winget Install Nushell) in my cmd
> 
>     * Then I installed the "vscode-nushell-lang" extension on Visual Studio Code
>       For now is this correct?
> 
> 
> Also send you a screen of the names of my workspace and strengthened according to the order I would like
> 
> ![Image](https://github.com/user-attachments/assets/4d8ed16d-1a37-4597-8cd6-13ea3a9c091d)
> 
> Thanks again for your work

Okay, I see, I mixed up an issue. I actually worked on this https://github.com/ferdium/ferdium-app/issues/1611.

However, since I have the base already there, it should not be difficult to apply it to workspaces, as well. I'll have a look.

**Comment by gabrielevinci** [comment]
ok, I remain waiting and I continue to follow then
Thanks for your work

**Comment by theAkito** [comment]
> ok, I remain waiting and I continue to follow then Thanks for your work

Try this. https://gist.github.com/theAkito/acd28326ae3c7e1732bcade8f5ec545e

**Comment by gabrielevinci** [comment]
> > ok, rimango in attesa e continuo a seguirti allora Grazie per il tuo lavoro
> 
> Prova questo. https://gist.github.com/theAkito/acd28326ae3c7e1732bcade8f5ec545e

I tried to run your file in my cmd, but it returned me a mistake. 
I attach the screen.
Ferdium obviously closed

![Image](https://github.com/user-attachments/assets/396c9c64-a15e-4fbf-aebf-85ee0145b96c)

**Comment by theAkito** [comment]
> > > ok, rimango in attesa e continuo a seguirti allora Grazie per il tuo lavoro
> > 
> > 
> > Prova questo. https://gist.github.com/theAkito/acd28326ae3c7e1732bcade8f5ec545e
> 
> I tried to run your file in my cmd, but it returned me a mistake. I attach the screen. Ferdium obviously closed
> 
> ![Image](https://github.com/user-attachments/assets/396c9c64-a15e-4fbf-aebf-85ee0145b96c)

As it says there, you can run the `--help` flag to get more information on how to run the script.

In your case, do this, for example.


```pwsh
nu ferdium-workspace-order-sqlite.nu "Personale"
```

**Comment by gabrielevinci** [comment]
> Nel tuo caso, fai questo, ad esempio.
> 
> nu ferdium-workspace-order-sqlite.nu "Personale"



done, but the cmd shows other errors
I continue to thank for the work you are doing for me and for the Ferdi community

![Image](https://github.com/user-attachments/assets/e70092c2-c1b7-4176-9ed2-f746987e4461)

**Comment by theAkito** [comment]
> > Nel tuo caso, fai questo, ad esempio.
> > nu ferdium-workspace-order-sqlite.nu "Personale"
> 
> done, but the cmd shows other errors I continue to thank for the work you are doing for me and for the Ferdi community
> 
> ![Image](https://github.com/user-attachments/assets/e70092c2-c1b7-4176-9ed2-f746987e4461)

https://gist.github.com/theAkito/daf71b45e02ecde502d55af4df0a6721

The second workspace related script is a generic one, which can work for anyone, as long as the configuration via YAML is set up correctly.

I've tested it thoroughly and works very well on my Ferdium client.

If you encounter issues with no workspaces or no services, just restore one of the backup SQLite databases in `C: \Users\User\AppataData\Roaming\Ferdium`.

**Comment by gabrielevinci** [comment]
I saw the code well and it is practically looking for the "workspace.yaml" file that does not exist on my PC. I also looked for it with "Everything" software but nothing. Maybe the file you are looking for is called in another way?

![Image](https://github.com/user-attachments/assets/e32ceb9f-4c9b-4202-adc5-11a0ea45ba85)

**Comment by theAkito** [comment]
> I saw the code well and it is practically looking for the "workspace.yaml" file that does not exist on my PC. I also looked for it with "Everything" software but nothing. Maybe the file you are looking for is called in another way?
> 
> ![Image](https://github.com/user-attachments/assets/e32ceb9f-4c9b-4202-adc5-11a0ea45ba85)

You have to create the file. It is a configuration file for my script.

##### `input/workspace.yaml`
```yaml
workspaces:
  Personal: 1
  Work: 2
  Project: 3
  Janitorial: 4
```

But replace "Personal", "Work" etc. with your workspace names, in the right order.

**Comment by gabrielevinci** [comment]
> > Ho visto bene il codice e sta praticamente cercando il file "workspace.yaml" che non esiste sul mio PC. L'ho cercato anche con un software "Everything", ma niente. Forse il file che stai cercando si chiama in un altro modo?
> > ![Image](https://github.com/user-attachments/assets/e32ceb9f-4c9b-4202-adc5-11a0ea45ba85)
> 
> Devi creare il file. È un file di configurazione per il mio script.
> 
> ##### `input/workspace.yaml`
> workspaces:
>   Personal: 1
>   Work: 2
>   Project: 3
>   Janitorial: 4
> Ma sostituisci "Personale", "Lavoro" ecc. con i nomi della tua area di lavoro, nel giusto ordine.




Ok, I created the file, but still I believe that the error of the code seems to come from fact that in my %ApptaData %\ Ferdium, the "server.sqlite" file does not contain anything inside it

![Image](https://github.com/user-attachments/assets/7c56b0d9-3d8d-4a88-b4fe-8cd38b6954d0)

**Comment by theAkito** [comment]
> > > Ho visto bene il codice e sta praticamente cercando il file "workspace.yaml" che non esiste sul mio PC. L'ho cercato anche con un software "Everything", ma niente. Forse il file che stai cercando si chiama in un altro modo?
> > > ![Image](https://github.com/user-attachments/assets/e32ceb9f-4c9b-4202-adc5-11a0ea45ba85)
> > 
> > Devi creare il file. È un file di configurazione per il mio script.
> > 
> > ##### `input/workspace.yaml`
> > workspaces:
> >   Personal: 1
> >   Work: 2
> >   Project: 3
> >   Janitorial: 4
> > Ma sostituisci "Personale", "Lavoro" ecc. con i nomi della tua area di lavoro, nel giusto ordine.
> 
> 
> 
> 
> Ok, I created the file, but still I believe that the error of the code seems to come from fact that in my %ApptaData %\ Ferdium, the "server.sqlite" file does not contain anything inside it
> 
> ![Image](https://github.com/user-attachments/assets/7c56b0d9-3d8d-4a88-b4fe-8cd38b6954d0)

If you open Ferdium and it works with an empty database, then I guess, the state is loaded from an online Ferdium account. However, I would expect some local state to be also present.

I'm still pretty new to Ferdium in general, so somebody else will probably have a better answer to that question.

All I can say for now, is that I did not log into any instance in my Ferdium client and have a whole SQLite database filled with the entire state of the workspaces, services, etc...

**Comment by gabrielevinci** [comment]
> Se apri Ferdium e funziona con un database vuoto, immagino che lo stato venga caricato da un account Ferdium online. Tuttavia, mi aspetterei che sia presente anche un po' di stato locale.
> 
> In generale sono ancora piuttosto nuovo per quanto riguarda Ferdium, quindi probabilmente qualcun altro saprà darmi una risposta migliore a questa domanda.
> 
> Tutto quello che posso dire per ora è che non ho effettuato l'accesso a nessuna istanza del mio client Ferdium e ho un intero database SQLite riempito con l'intero stato degli spazi di lavoro, dei servizi, ecc...

I managed to make the code work. A thousand thanks.
To make myself useful, I will share a step-by-step guide for all beginners (like me) who want to do the same thing but understand nothing about it.

Before starting:
- This guide is used to order Ferdium's worksplas, not the services.
- This guide works only if you use Ferdium without login in account (it makes you choose the first opening after installation)
- If you have used your account by now, come out of it, recreate workspace and services and then implement the guide

guida step by step:

1. Open the cmd and type "Winget Install Nushell", press Send (if you don't have Winget install it)
2. Create a folder in your desktop (random name, in this guide we will call it "tests")
3. Inside the "Test" folder, create a TXT file but glue this name "ferdium-workspace-order-sqlite.nu", will ask you if you want to change the file format, you press on "yes"
4. Open this file with "notes block" and glue the second code of this page (https://gist.github.com/theAkito/daf71b45e02ecde502d55af4df0a6721#file-ferdium-service-order-sqlite-nu), the code called "ferdium-workspace-order-sqlite.nu", kindly offered by @theAkito 
5. In the "test" folder, create a folder and call it "input"
6. Inside the input folder, create a TXT file and call it "workspace.yaml", change the file format here too.
7. Open this file and you will have to write how you want to order your workspaces, in this format:

```
workspaces:
  Personal: 1
  Work: 2
  Project: 3
  Janitorial: 4
```
Instead of "Personal/Work .... etc." you will have to put the names of your worksplas in the order you want- salvate il file e chiudetelo

- Close Ferdium completely (not by pressing on the "X" at the top right, but by pressing Ctrl+Q)
- Open the "Test" folder
- Click in the part of the folder where the path is indicated (where you will see "tests"). If you click there, such a thing in blue "C:\Users\Virtu\Desktop\Test" ("virtu" changes from computer to computer) will emerge.
- write "cmd" and press send
- In the CMD, glue this command: nu ferdium-workspace-order-sqlite.nu
- Press send
- the game is done

A screen like the one you see in the photo should appear to you

I hope this guide has been helpful and we thank you again @theAkito 

![Image](https://github.com/user-attachments/assets/91e8316c-d361-4946-abb3-f20e4ff5a522)

**Comment by theAkito** [comment]
> > Se apri Ferdium e funziona con un database vuoto, immagino che lo stato venga caricato da un account Ferdium online. Tuttavia, mi aspetterei che sia presente anche un po' di stato locale.
> > In generale sono ancora piuttosto nuovo per quanto riguarda Ferdium, quindi probabilmente qualcun altro saprà darmi una risposta migliore a questa domanda.
> > Tutto quello che posso dire per ora è che non ho effettuato l'accesso a nessuna istanza del mio client Ferdium e ho un intero database SQLite riempito con l'intero stato degli spazi di lavoro, dei servizi, ecc...
> 
> Sono riuscito a far funzionare il codice. grazie mille. Per rendermi utile, condividerò una guida step-by-step per tutti i principianti (come me) che vogliono fare la stessa cosa ma non ne capiscono nulla.
> 
> Prima di iniziare:
> 
>     * Questa guida serve per ordinare i workspace di ferdium, non i servizi.
> 
>     * Questa guida funziona solo se usate ferdium senza login in account (ve lo fa scegliere alla prima apertura dopo l'installazione)
> 
>     * se ormai avete usato il vostro account, uscite da esso, ricreate workspace e servizi e poi mettete in atto la guida
> 
> 
> guida step by step:
> 
>     * aprite il cmd e digitate "Winget Install Nushell", premete invio (se non avete winget installatelo)
> 
>     * create una cartella nel vostro desktop (nome casuale, in questa guida la chiameremo "test")
> 
>     * dentro la cartella "test" create un file txt ma incollate questo nome "ferdium-workspace-order-sqlite.nu", vi chiederà se vorrete cambiare il formato del file, voi premete su "si"
> 
>     * aprite questo file con "blocco appunti" e incollate il secondo codice di questa pagina (https://gist.github.com/theAkito/daf71b45e02ecde502d55af4df0a6721#file-ferdium-service-order-sqlite-nu), il codice che si chiama "ferdium-workspace-order-sqlite.nu", gentilmente offerto da [@theAkito](https://github.com/theAkito)
> 
>     * nella cartella "test" create una cartella e la chiamate "input"
> 
>     * dentro la cartella input create un file txt e lo chiamate "workspace.yaml", cambiate anche qui il formato del file.
> 
>     * aprite questo file e dovrete scrivere come vorrete ordinare i vostri workspace, in questo formato:
> 
> 
> ```
> workspaces:
>   Personal: 1
>   Work: 2
>   Project: 3
>   Janitorial: 4
> ```
> 
> al posto di "Personal/Work....ecc" dovrete mettere i nomi dei vostri workspace nell'ordine che desiderate
> 
>     * salvate il file e chiudetelo
> 
>     * chiudete completamente ferdium (non premendo sulla "X" in alto a destra, ma premendo ctrl+q)
> 
>     * aprite la cartella "test"
> 
>     * cliccate nella parte della cartella in cui è indicato il percorso (dove vedrete "test"). se ci cliccate vi spunterà una cosa del genere in blu "C:\Users\virtu\Desktop\test" ("virtu" cambia da computer a computer).
> 
>     * scrivete "cmd" e premete invio
> 
>     * nel cmd incollate questo comando: nu ferdium-workspace-order-sqlite.nu
> 
>     * premete invio
> 
>     * il gioco è fatto
> 
> 
> vi dovrebbe apparire una schermata come quella che vedete in foto
> 
> spero che questa guida vi sia stata d'aiuto e ringraziamo ancora [@theAkito](https://github.com/theAkito)
> 
> ![Image](https://github.com/user-attachments/assets/91e8316c-d361-4946-abb3-f20e4ff5a522)

Hey buddy, as far as I understand, my work helped you. I'm glad!

You might want to check your auto-translator though, as this message you posted isn't translated.

I would also recommend to post your guide via Github Gist, for example, to make it easily shareable and editable. 🙂

**Comment by gabrielevinci** [comment]
> Hey buddy, as far as I understand, my work helped you. I'm glad!
> 
> You might want to check your auto-translator though, as this message you posted isn't translated.
> 
> I would also recommend to post your guide via Github Gist, for example, to make it easily shareable and editable. 🙂



I had not translated the entire message for fear that the translation could have distorted commands/links ... etc., in addition to thinking that everyone has Google Traslator integrated on their PC, so everyone could have translated the guide into his language. However, I have translated it slowly for coherence, I stand up not to distort commands or links. 

I would like to follow your advice to put it on Github but I'm brand new in here, these are my very first posts, so maybe I should look at me some tutorials to understand how to create it and integrate it somehow here

**Comment by theAkito** [comment]
Just FYI: I extended my workspace & service manipulation project and corrected the code. The original Gist had a major bug.

https://github.com/theAkito/ferdium-tool

**Comment by Lewis8265** [comment]
I'm absolutely baffled why this isn't default behaviour. Very strange. 
For me, this is the app's biggest weakness that is enough to put me off of it.
Even automatic alphabetical sorting would be fine as at least I could prefix a letter to make things go in a sensible order :)

**Comment by KB8DOA** [comment]
I was quite excited about Ferdium until I ran into this same issue.

Not being able to **customize the sort of the Services** is a major problem for me.
This is quite an obvious issue, and I don't get why something like this could have been fixed.

I have much respect for what has been achieved with Ferdium, but have major concern about it's future
when issues like this have been made aware, and discussed, but still not fixed...
